app/stable_appconfig.go

```diff
@@ -59,7 +68,102 @@ func setCustomOrderEndBlocker() []string {
 	return []string{}
 }
 
-func (app *RegenApp) registerUpgradeHandlers() {}
+func (app *RegenApp) registerUpgradeHandlers() {
+
+	// mainnet upgrade handler
+	const upgradeName = "v4.0.0"
+	app.UpgradeKeeper.SetUpgradeHandler(upgradeName, func(ctx sdk.Context, plan upgradetypes.Plan, fromVM module.VersionMap) (module.VersionMap, error) {
+		// run state migrations for sdk modules
+		toVersion, err := app.mm.RunMigrations(ctx, app.configurator, fromVM)
+		if err != nil {
+			return nil, err
+		}
+
+		// run x/ecocredit state migrations
+		if err := app.smm.RunMigrations(ctx, app.AppCodec()); err != nil {
+			return nil, err
+		}
+		toVersion[ecocredit.ModuleName] = ecocreditmodule.Module{}.ConsensusVersion()
+
+		if err := recoverFunds(ctx, app.AccountKeeper, app.BankKeeper); err != nil {
+			return nil, err
+		}
```

Comment:
Separate upgrade handlers are no longer needed if we are performing manual operations in the state migrations based on chain id. We should check chain id here and run recover funds only for chain id `regen-1` and remove the separate upgrade handler for redwood so that this upgrade handler works for `regen-1`, `regen-redwood-1`, and for any other test networks with a separate chain id. Apologies for the confusion.

```suggestion
		// recover funds for community member (regen-1 governance proposal #11)
		if ctx.ChainID() == "regen-1" {
			if err := recoverFunds(ctx, app.AccountKeeper, app.BankKeeper); err != nil {
				return nil, err
			}
		}
```